app: esphome

FeatherKing opened the pull request:
Adding ESPHome. Their logo is located [here](https://esphome.io/images/index.html). let me know what the new links are and i will update the PR.

The ESPHome version in this pr is 2024.12. host networking is disabled. ESPHome relies on mDNS for device discovery so without host networking device discovery wont work. I can add that in a description somewhere if needed. ESPHome documents this behavior [here](https://esphome.io/guides/faq.html#docker-reference-notes). Tests are passing and I can get to the UI. I didnt actually try to deploy a device but it seemed like i could get decently far.

domoritz commented on ix-dev/community/esphome/README.md:
```diff
@@ -0,0 +1,3 @@
+# esphome
+
+[esphome](https://github.com/esphome/esphome) is a PVR for Usenet and BitTorrent users. It can monitor multiple RSS feeds for new episodes of your favorite shows and will grab, sort and rename them.
```
That's not the description for esphome.

FeatherKing replied:
fixed in [182bc346](https://github.com/truenas/apps/pull/1406/commits/182bc34606d189dc253d89a4a27be5dfe78ba827)

domoritz commented on ix-dev/community/esphome/item.yaml:
```diff
@@ -0,0 +1,7 @@
+categories:
+- productivity
```
Is that the right category?

FeatherKing replied:
i copied this from the home assistant app. Is there a list of categories somewhere?

domoritz replied:
I don't know. This just stood out. I asked in https://github.com/truenas/apps/discussions/1409.

domoritz replied:
I found https://github.com/truenas/apps/blob/master/ix-dev/community/zigbee2mqtt/item.yaml#L2C3-L2C18

```suggestion
- home-automation
```

domoritz commented on ix-dev/community/esphome/ix_values.yaml:
```diff
@@ -0,0 +1,9 @@
+images:
+  image:
+    repository: ghcr.io/esphome/esphome
```
Looks good overall but this image won't support avahi out of the box, right?

FeatherKing replied:
out of the box, no. Upstream esphome doesnt have it and recommends the environment variable ive used here to enable ping checks. However, ive seen some other builds like [this ](https://github.com/vxav/esphome-avahi) that add avahi to the upstream image. Im not sure if this is that maintainable (this image is already a year old).

domoritz replied:
Not required but would be nice. I found that the ping doesn't work when the dns doesn't resolve. Happy to test the current version when it's available, though.

stavros-k commented on ix-dev/community/esphome/questions.yaml:
```diff
@@ -0,0 +1,386 @@
+    schema:
+      type: dict
+      attrs:
+        - variable: instance_name
+          label: Instance Name
+          description: The name of the esphome instance.
+          schema:
+            type: string
+            default: "esphome"
+            required: true
```
Where is this used?

FeatherKing replied:
removed. left over copy/paste

stavros-k commented on ix-dev/community/esphome/questions.yaml:
```diff
@@ -0,0 +1,386 @@
+            default: 36052
+            required: true
+            $ref:
+              - definitions/port8
```
port8?

FeatherKing replied:
fat finger, removed

stavros-k commented on ix-dev/community/esphome/templates/docker-compose.yaml:
```diff
@@ -0,0 +1,22 @@
+{% do c1.environment.add_env("ESPHOME_DASHBOARD_USE_PING", true) %}
+{% do c1.environment.add_user_envs(values.esphome.additional_envs) %}
+
+{% do c1.ports.add_port(values.network.web_port, values.consts.internal_webui_port) %}
```
```suggestion
{% do c1.add_port(values.network.web_port, {"container_port": values.consts.internal_webui_port}) %}
```
Please see latest apps added like zigbee2mqtt how the new port is added and how the questions.yaml looks like.

stavros-k commented on ix-dev/community/esphome/app.yaml:
```diff
@@ -0,0 +1,31 @@
+  url: https://www.truenas.com/
+name: esphome
+run_as_context:
+- description: esphome runs as any non-root user.
```
```suggestion
- description: ESPHome runs as any non-root user.
```

FeatherKing replied:
updated

stavros-k commented on ix-dev/community/esphome/app.yaml:
```diff
@@ -0,0 +1,31 @@
+- https://media.sys.truenas.net/apps/esphome/screenshots/screenshot1.png
+sources:
+- https://github.com/esphome/esphome
+title: esphome
```
```suggestion
title: ESPHome
```

FeatherKing replied:
updated

stavros-k commented on ix-dev/community/esphome/templates/docker-compose.yaml:
```diff
@@ -0,0 +1,22 @@
+{% do c1.environment.add_env("ESPHOME_DASHBOARD_USE_PING", true) %}
+{% do c1.environment.add_user_envs(values.esphome.additional_envs) %}
+
+{% do c1.ports.add_port(values.network.web_port, values.consts.internal_webui_port) %}
```
Is internal port configurable? 
That would make much easier to allow enabling host network

FeatherKing replied:
i looked [here ](https://github.com/esphome/esphome/blob/dev/docker/Dockerfile)and it seems hardcoded to 6052. Also found this [feature request](https://github.com/esphome/feature-requests/issues/2738) for the same

stavros-k commented on ix-dev/community/esphome/questions.yaml:
```diff
@@ -0,0 +1,386 @@
+groups:
+  - name: esphome Configuration
```
```suggestion
  - name: ESPHome Configuration
```
Please rename everywhere here

FeatherKing replied:
updated

stavros-k commented on ix-dev/community/esphome/README.md:
```diff
@@ -0,0 +1,3 @@
+# esphome
+
+[esphome](https://github.com/esphome/esphome) is a system to control your microcontrollers by simple yet powerful configuration files and control them remotely through Home Automation systems.
```
```suggestion
# ESPHome

[ESPHome](https://github.com/esphome/esphome) is a system to control your microcontrollers by simple yet powerful configuration files and control them remotely through Home Automation systems.
```

FeatherKing replied:
renamed

FeatherKing commented:
@stavros-k ive addressed all the feedback, thanks for the first review.
How should i select the host port in the future? i used 36052 here but that was just adding a 3 to the front of their internal port 6052. I also couldnt address the comment on changing the internal port, to me it doesnt seem like it can be configured.

My tests are passing, ill wait for your next feedback. Also let me know what to do with screenshot links.

stavros-k commented:
> @stavros-k ive addressed all the feedback, thanks for the first review. How should i select the host port in the future? i used 36052 here but that was just adding a 3 to the front of their internal port 6052. I also couldnt address the comment on changing the internal port, to me it doesnt seem like it can be configured.
> 
> My tests are passing, ill wait for your next feedback. Also let me know what to do with screenshot links.

I usually pick a port that is not already used by other app. Its just to avoid conflicts if someone tries to install all the things lol.

stavros-k commented:
I'll handle icon/screenshots tomorrow!

stavros-k commented on ix-dev/community/esphome/questions.yaml:
```diff
@@ -0,0 +1,402 @@
+                label: Port Number
+                schema:
+                  type: int
+                  show_if: [["bind_mode", "!=", ""]]
```
```suggestion
                  show_if: [["bind_mode", "=", "published"]]
```
Since internal port is not changable. Only this mode should allow for setting port number
The whole web_port should show_if host_network is false.

stavros-k reviewed (approved):
Thanks

domoritz commented:
Thank you. I think the container needs privileged, though. See https://github.com/truenas/apps/issues/1451.

I wasn't able to compile a spec. mdns also doesn't seem to be working yet even with host mode on.

FeatherKing commented:
privileged doesnt solve it, but running as root lets me compile. Looking to see if there is a way around this error when compiling in the container

> PermissionError: [Errno 13] Permission denied: '/.platformio'

FeatherKing commented:
> Thank you. I think the container needs privileged, though. See #1451.
> 
> I wasn't able to compile a spec. mdns also doesn't seem to be working yet even with host mode on.

made https://github.com/truenas/apps/pull/1453
I was able to compile for several device types. I think there is still a better way to do it, but this is working. I took a yaml from my existing esphome and compiled a new build with this for one of my light switches.

mdns wont work unless we use an avahi build, but this esphome was able to connect to a light switch which is on another vlan and show its logs in the web ui.
